rs/node_rewards/canister/src/registry.rs

```diff
@@ -0,0 +1,430 @@
+use ic_nervous_system_canisters::registry::Registry;
+use ic_protobuf::registry::dc::v1::DataCenterRecord;
+use ic_protobuf::registry::node::v1::{NodeRecord, NodeRewardType};
+use ic_protobuf::registry::node_operator::v1::NodeOperatorRecord;
+use ic_protobuf::registry::node_rewards::v2::NodeRewardsTable;
+use ic_protobuf::registry::subnet::v1::SubnetListRecord;
+use ic_registry_canister_client::{
+    CanisterRegistryClient, RegistryDataStableMemory, StableCanisterRegistryClient,
+    StorableRegistryKey, StorableRegistryValue,
+};
+use ic_registry_keys::{
+    make_subnet_list_record_key, DATA_CENTER_KEY_PREFIX, NODE_OPERATOR_RECORD_KEY_PREFIX,
+    NODE_RECORD_KEY_PREFIX, NODE_REWARDS_TABLE_KEY,
+};
+use ic_types::registry::RegistryClientError;
+use indexmap::IndexMap;
+use rewards_calculation::rewards_calculator_results::DayUTC;
+use rewards_calculation::types::{
+    NodeType, ProviderRewardableNodes, Region, RewardableNode, UnixTsNanos,
+};
+use std::collections::{BTreeMap, HashMap};
+use std::str::FromStr;
+use std::sync::Arc;
+
+pub trait RegistryEntry: RegistryValue {
+    const KEY_PREFIX: &'static str;
+}
+
+impl RegistryEntry for DataCenterRecord {
+    const KEY_PREFIX: &'static str = DATA_CENTER_KEY_PREFIX;
+}
+
+impl RegistryEntry for NodeOperatorRecord {
+    const KEY_PREFIX: &'static str = NODE_OPERATOR_RECORD_KEY_PREFIX;
+}
+
+impl RegistryEntry for NodeRecord {
+    const KEY_PREFIX: &'static str = NODE_RECORD_KEY_PREFIX;
+}
+
+impl RegistryEntry for NodeRewardsTable {
+    const KEY_PREFIX: &'static str = NODE_REWARDS_TABLE_KEY;
+}
+
+pub struct RegistryClient<S: RegistryDataStableMemory> {
+    store: Arc<StableCanisterRegistryClient<S>>,
+}
+
+struct NodeOperatorData {
+    node_provider_id: PrincipalId,
+    dc_id: String,
+    region: Region,
+}
+
+enum RegistryRecordState {
+    Present(UnixTsNanos),
+    Deleted(UnixTsNanos),
+}
+
+struct RegistryRecordWithStates {
+    key: String,
+    value: Vec<u8>,
+    registry_record_states: Vec<RegistryRecordState>,
+}
+
+impl<S: RegistryDataStableMemory> RegistryClient<S> {
+    pub fn new(registry: Arc<dyn Registry>) -> Self {
+        let store = Arc::new(StableCanisterRegistryClient::<S>::new(registry));
+        RegistryClient { store }
+    }
+
+    pub fn store(&self) -> Arc<StableCanisterRegistryClient<S>> {
+        self.store.clone()
+    }
+
+    pub async fn sync_registry_stored(&self) -> Result<RegistryVersion, String> {
+        self.store.sync_registry_stored().await
+    }
+
+    pub fn get_versioned_value<T: RegistryValue + Default>(
+        &self,
+        key: &str,
+        version: RegistryVersion,
+    ) -> Result<T, RegistryClientError> {
+        let value = self
+            .store
+            .get_versioned_value(key, version)?
+            .map(|v| T::decode(v.as_slice()).unwrap())
+            .value
+            .unwrap_or_default();
+        Ok(value)
+    }
+
+    pub fn get_value<T: RegistryValue + Default>(
```

Comment:
you could drop this and use the store impl where it's used (since it's only in this file)